SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp

```diff
@@ -0,0 +1,153 @@
+    source_simd.copy_from(ref_data);
+    simd<DataT, NumElems> simd_by_var_decl{source_simd};
+    return simd_by_var_decl;
+  }
+};
+
+// Struct that provide method that let get current test case description and
+// method that let call simd constructor for rvalue in an expression test case
+struct rval_in_expression {
+  static std::string get_description() { return "rvalue in an expression"; }
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems> call_simd_ctor(const DataT *ref_data) {
+    simd<DataT, NumElems> source_simd{};
+    source_simd.copy_from(ref_data);
+    simd<DataT, NumElems> simd_by_rval;
+    simd_by_rval = simd<DataT, NumElems>(source_simd);
+    return simd_by_rval;
+  }
+};
+
+// Struct that provide method that let get current test case description and
+// method that let call simd constructor for const reference test case
+struct const_ref {
+  static std::string get_description() { return "const reference"; }
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems>
+  call_simd_by_const_ref(const simd<DataT, NumElems> &simd_by_const_ref) {
+    return simd_by_const_ref;
+  }
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems> call_simd_ctor(const DataT *ref_data) {
+    simd<DataT, NumElems> source_simd{};
+    source_simd.copy_from(ref_data);
+    return call_simd_by_const_ref<DataT, NumElems>(
+        simd<DataT, NumElems>(source_simd));
+  }
+};
+
+// The main test routine.
+// Use struct with overloaded call operator to have able for iterating over
+// pre-determined datatypes.
+template <typename DataT, int NumElems, typename TestCaseT> struct test {
+  bool operator()(sycl::queue &queue, const std::string &data_type) {
+    bool passed{true};
+
+    vector_with_allocator<DataT> ref_data{allocator_type<DataT>{queue}};
+    generate_ref_data<DataT, NumElems>(ref_data);
+
+    auto result_data = call_simd<DataT, NumElems, TestCaseT>(queue, ref_data);
+
+    for (size_t it = 0; it < ref_data.size(); it++) {
+      if (ref_data[it] !=
+          result_data[it]) { // here should be bitwise comparison
+        passed = false;
+        log::fail<NumElems>(
+            "Simd by " + TestCaseT::get_description() +
+                " failed, retrieved: " + std::to_string(result_data[it]) +
+                ", expected: " + std::to_string(ref_data[it]),
+            data_type);
+      }
+    }
+
+    return passed;
+  }
+};
+
+template <typename DataT, typename TestT>
+using run_test_with_one_elem = test<DataT, 1, TestT>;
+
+template <typename DataT, typename TestT>
+using run_test_with_eight_elems = test<DataT, 8, TestT>;
+
+template <typename DataT, typename TestT>
+using run_test_with_sixteen_elems = test<DataT, 16, TestT>;
+
+template <typename DataT, typename TestT>
+using run_test_with_thirty_two_elems = test<DataT, 32, TestT>;
+
+template <typename TestT, typename... T>
+bool run_verification_with_chosen_test_type(
```

Comment:
What about something like `for_all_types_and_dims(named_type_pack, integer_pack)` so we don't need all these wrappers?
We might strictly need it for other tests - still looks like out of scope of this PR though)

Reply:
I can done it while I will implementing tests for fill constructor